### DISCUSSION TITLE
Passing an arbitrary string as option

### DISCUSSION THREAD

**Discussion by Max13** [Homebrew/discussions#2305] [Writing Formulae/Casks]
I can't find how to pass an arbitrary string as an option (cf: homebrew/brew#1803), apparently, `ARGV` can't be used anymore.

I'm writing a formula that depends on `Qt`, and the user should be able to provide his own installation instead of installing the runtime by `homebrew`.

Thanks for your help

**Comment by SMillerDev** [comment]
> I'm writing a formula that depends on Qt, and the user should be able to provide his own installation instead of installing the runtime by homebrew.

Asside from the option string issue you're encounter, almost everything in brew is working to prevent brew from doing this because it makes builds really unreliable.

**Reply by Max13** [reply]
> almost everything in brew is working to prevent brew from doing this because it makes builds really unreliable.

I understand, but when you need my lib being built against the current installed `Qt`, to use it then with an app you are writing with `Qt`, then how would the user indicate his currently installed `Qt` instead of reinstalling one from `Homebrew`?

**Reply by SMillerDev** [reply]
Well, if you don't want homebrew to manage the packages I would just avoid homebrew altogether.

**Reply by Max13** [reply]
So I understand there is no way? I'm using brew for the easy build process, once the formula is written, it's (virtually) frozen. Is there a workaround?

I have a formula with either `Qt` or `Boost`, but there is no sense in letting brew build my libs with a formula while the final app will be bound with another 😕

If you confirm there is **no way**, then I won't insist.

Thanks

**Reply by SMillerDev** [reply]
I'm not a hundred percent sure there is no way. But I do know that this is something we try very hard to avoid in homebrew because it makes building software reliably very difficult.

**Reply by carlocab** [reply]
> I have a formula with either `Qt` or `Boost`, but there is no sense in letting brew build my libs with a formula while the final app will be bound with another 😕

You can use `:optional` and/or `:recommended`, for this. For example:

```ruby
depends_on "boost" => :recommended
depends_on "qt" => :optional
```

This will build your formula with `boost` and without `qt`. To do the reverse, a user can pass `brew install --with-qt --without-boost`. You can probably make the `--without-boost` implied by `--with-qt` if you make judicious use of `build.with?("qt")`.

If you want users to build with a custom version of Qt, one thing that might work is to ask users to symlink their Qt installation prefix into `/tmp/custom-qt` and then add that path to the include and library search path using `C[PP,XX]FLAGS`, which you can probably gate behind a custom option (e.g. `--with-custom-qt`). Not 100% sure this will work, though.
